academy/ibc-dev/ibcapp_checkers.md

```diff
@@ -0,0 +1,342 @@
+
+...
+
+// ValidateBasic is used for validating the packet
+func (p IbcTopRankPacketData) ValidateBasic() error {
+
+  // TODO: Validate the packet data
+  
+  // check the gameID
+  if p.GameID != p.ChannelID {
+    return errors.New("GameID does not match")
+  }
+
+  return nil
+}
+
+...
+
+```
+
+this way we make sure that we do not give any other chain the permission to alter the wrong leaderboard by sending a wrong `GameID`. 
+
+## Test with Docker
+
+You can find a test network in the [sample](https://github.com/b9lab/cosmos-ibc-docker/tree/main/separate). First you will need clone the [sample](https://github.com/b9lab/cosmos-ibc-docker/tree/main/separate) and build the docker image for the `checkers` chain:
+
+```bash
+$ cd cosmos-ibc-docker/separate/checkers
+$ docker build -f Dockerfile . -t checkers --build-arg configyml=./config.yml --no-cache
+```
+
+and then the image for the the `leaderboard` chain:
+
+```bash
+$ cd cosmos-ibc-docker/separate/leaderboard
+$ docker build -f Dockerfile . -t leaderboard --build-arg configyml=./config.yml --no-cache
+```
+
+as well as the image for the relayer:
+
+```bash
+$ cd cosmos-ibc-docker/separate/relayer
+$ docker build -f Dockerfile . -t relayer --no-cache
+```
+
+You can use the provided compose file to spin up a network with a `checkers` blockchain and a `leaderboard` blockchain as well as a relayer to relay between them:
+
+```bash
+$ cd cosmos-ibc-docker/separate
+$ docker-compose -f separate.yml up
+
+```
+
+Observe the output of `docker-compose` until the chains are ready - it will take some time for the chains to be ready. If the chains are ready, you can start the relayer process with:
+
+```bash
+$ docker exec relayer ./run-relayer.sh 
+```
+
+wait till the connection is etablished and a channel is created. 
+
+Then jump into checkers container:
+
+```bash
+$ docker exec -it checkers bash
+```
+
+let the test script create and play a game:
+
+```bash
+$ ./test.sh 1 checkers cosmos1n4mqetruv26lm2frkjah86h642ts84qyd5uvyz cosmos14y0kdvznkssdtal2r60a8us266n0mm97r2xju8
+```
+
+You can use this script many times but make sure to increase the first argument, which indicates the game index.
+
+This again will take some time and you will see transaction flying around. If the game is over, you can query the local leaderboard of the `checkers` chain with:
+
+```bash
+$ checkersd q checkers show-leaderboard
+```
+
+Now you can send the score via:
+
+```bash
+$ checkersd tx checkers send-score checkers channel-0 --from cosmos1n4mqetruv26lm2frkjah86h642ts84qyd5uvyz --chain-id checkers
+```
+
+Note that only the winner of the game will be in the local leaderboard of the `checkers` chain so there is no score for the loser to send yet. 
+
+After a short time, you can find a new entry in the leaderboard of the `leaderboard` chain. Therefore jump into leaderboard chain:
+
+```bash
+$ docker exec -it leaderboard bash
+```
+
+and check the board:
+
+```bash
+$ leaderboardd q leaderboard list-board
+```
```

Comment:
Looks incomplete but I have not run the project to know what it will output.

Reply:
You mean it would be better to have the output included?